### ISSUE TITLE
OC 3.0.1.2: Clashes between theme editor and view/*/before events

### ISSUE THREAD

**Issue by mhcwebdesign** [opencart/opencart#5707]
When using the OpenCart theme editor at the admin backend at

Design > Theme Editor

to e.g. edit the common/header.twig, and some 3rd party extensions use event handlers for the triggers **catalog/view/common/header/before** then the latter aren't being executed. The chain of triggering events is interrupted prematurely. This makes using event handlers for modifying templates pretty useless. Not even OCmod will get a chance to do its modification in case of an above mentioned theme override. 

Basically, the theme editor stores its changes to template files in the database, in table 'oc_theme'. And we need a mechanism for these modified theme files to be subjected to subsequent further modifications done via OCmod or via 'catalog/view/*/before' event handlers.

**Comment by danielkerr** [comment]
as long as the ControllerEventTheme is called last it shoudl work right?

**Comment by mhcwebdesign** [comment]
But the ControllerEventTheme is not called last! 

First it always executes the 'system/startup.php' where it loads the events from the 'system/config/catalog.php', namely the:

event/language/before
event/language/after
event/language
event/theme
event/translation

Then, later on, it loads the events from the database at catalog/controller/startup/event.php, including user defined events for e.g. the trigger 'view/common/header/before'.

If you take a look at the 'system/engine/event.php', in method register($trigger, Action $action, $priority = 0), it has indeed a 'priority' argument, but it doesn't use it! Hence the events for a given trigger are executed in the order in which they were registered chronologically, not according to the sort_order (that is priority)!

**Comment by danielkerr** [comment]
let me know if this fixes it.

**Comment by danielkerr** [comment]
im going to do some testing now.

**Comment by danielkerr** [comment]
ok so if i do 

```
'view/common/footer/before' => array(
	'event/debug/before'
),
```

and add into the event i get
```
	public function before($route, &$args) {
		//if ($route == 'common/home') { // add the route you want to test
		//	$this->session->data['debug'][$route] = microtime();
		//}
		
		//print_r($args);
		
		$args['text_contact'] = 'test';
		
	}
```

it changes the text to footer contact us text to test

**Comment by danielkerr** [comment]
ok checking the theme

**Comment by mhcwebdesign** [comment]
Thanks, I'll test soon. 

As long as the event triggers take the priority into account, too, it should work.

**Comment by danielkerr** [comment]
just tested the theme editor and even though on the front side the events are in order its still not showing the changed data

ControllerEventTranslation
ControllerEventLanguage
ControllerEventTheme

**Comment by mhcwebdesign** [comment]
I just tested it. I have a module which adds a frontend event for trigger 'view/common/header/before', and it's not even triggered. The 'view/*/before' are always listed first before e.g. the 'view/common/header/before'. 

They really ought be listed by priority first, e.g. in the **system/event.php** it should use something like this:
````
	public function register($trigger, Action $action, $priority = 0) {
		if (!isset($this->data[$priority])) {
			$this->data[$priority] = array();
		}
		if (isset($this->data[$priority][$trigger])) {
			$this->data[$priority][$trigger] = array();
		}
		$this->data[$priority][$trigger][] = $action;
	}
	
	public function trigger($event, array $args = array()) {
		foreach ($this->data as $priority=>$triggers) {
			foreach ($triggers as $trigger=>$actions) {
				if (preg_match('/^' . str_replace(array('\*', '\?'), array('.*', '.'), preg_quote($trigger, '/')) . '/', $event)) {
					foreach ($actions as $action) {
						$result = $action->execute($this->registry, $args);
						if (!is_null($result) && !($result instanceof Exception)) {
							return $result;
						}
					}
				}
			}
		}
	}
````
 I haven't yet tested this approach, this is just a quick thought, but it should solve it.

**Comment by danielkerr** [comment]
it is but when there are multiple calls it breaks. coming upo with a fix now.

```
	'view/*/before' => array(
		999  => 'event/language',
		1000 => 'event/theme'
	),
	'view/common/footer/before' => array(
		900 => 'event/debug/before'
	)
```

`1000 => 'event/theme'` is called 2nd and because it returns a value then `900 => 'event/debug/before'` is not loaded.

**Comment by danielkerr** [comment]
good point!

**Comment by mhcwebdesign** [comment]
If there are 2 actions with same priority and same trigger, they'd just be added both in the array, e.g.:

$this->data[2]['view/common/header/before'][0] = action1
$this->data[2]['view/common/header/before'][1] = action2

Take a look at my suggested code!

**Comment by danielkerr** [comment]
managed to get it working but very strange behavior. i have to use ksort to get the array to sort properly.

**Comment by mhcwebdesign** [comment]
There should be no need for a ksort if you use an array indexed by:
- priority
- trigger
- default integer index (as per the chronological the actions are registered)

**Comment by danielkerr** [comment]
try the latest commit
